report_data.py

```diff
@@ -235,19 +308,21 @@ def process_sales(self, df_sales):
             # to compute the "Kapitalertrag"
             self.add_fees(row, f"Selling {row.symbol}")
 
-            currency, new_forex = FIFOForex.from_share_sale(row)
+            _, new_forex = FIFOForex.from_share_sale(row)
             self.held_forex[row.currency].push(new_forex)
 
     def process_forex_to_eur(self, df_forex_to_eur):
         # When doing a currency conversion, you convert the USD you possess into the equivalent amount of EUR.
         # This doesn't include the fee you pay for the transfer, that just vanishes in the original denomination
-        for row_idx, row in df_forex_to_eur.iterrows():
+        for _, row in df_forex_to_eur.iterrows():
             sold_currency = row.currency
-            self.held_forex[sold_currency].pop(row.fees, row.date)  # remove fees
-            tmp = self.held_forex[sold_currency].pop(row.net_amount, row.date)
+            fees = to_decimal(row.fees)
+            net_amount = to_decimal(row.net_amount)
+            self.held_forex[sold_currency].pop(fees, row.date)  # remove fees
+            tmp = self.held_forex[sold_currency].pop(net_amount, row.date)
             for t in tmp:
                 t.sell_date = row.date
-                t.sell_price = 1  # currency unit
+                t.sell_price = to_decimal(1)  # currency unit
```

Comment:
I don't know too much about python, but in C++ there's a concept of "smart structures" which take care of these accidental "I forgot to convert this to Decimal" mistakes. Within the structure itself the setter takes care of the conversion and you, as the user of the structure don't need to care about the inner logic.

Do you think it's possible to implement (safe-guard) something like this in python?

Reply:
This will be quite hard with the dynamic nature of Python I guess. There are reasons why people use things like `@dataclass(frozen=True)` but most of the times this is also not really perfect.